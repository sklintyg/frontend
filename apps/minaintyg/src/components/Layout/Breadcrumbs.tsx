import { IDSBreadcrumbs, IDSCrumb } from '@frontend/ids-react-ts'
import { Link, Params, useMatches } from 'react-router-dom'

type Match = ReturnType<typeof useMatches>[number]
type MatchWithCrumb = Match & {
  handle: { crumb: (data: Params<string>) => string }
}

function hasCrumb(match: MatchWithCrumb | Match): match is MatchWithCrumb {
  return (match as MatchWithCrumb).handle !== undefined && typeof (match as MatchWithCrumb).handle.crumb === 'function'
}

function getMatchAt(index: number, matches: MatchWithCrumb[]): { pathname: string; crumb: string } | null {
  const match = matches.at(index)
  return match ? { pathname: match.pathname, crumb: match.handle.crumb(match.params) } : null
}

export function Breadcrumbs() {
  const matches = useMatches().filter(hasCrumb)
  const prevMatch = getMatchAt(-2, matches)
  const currentMatch = getMatchAt(-1, matches)

  return (
    <div className="mb-5">
      <IDSBreadcrumbs current={currentMatch?.crumb ?? 'Start'} srlabel="Du 채r h채r" lead="Du 채r h채r:">
        {matches.slice(0, -1).map(({ handle, params, pathname }) => (
          <IDSCrumb key={pathname}>
            <Link to={pathname}>{handle.crumb(params)}</Link>
          </IDSCrumb>
        ))}
        {prevMatch && (
          <IDSCrumb key="mobile" mobile>
            <Link to={prevMatch.pathname}>{prevMatch.crumb}</Link>
          </IDSCrumb>
        )}
      </IDSBreadcrumbs>
    </div>
  )
}
