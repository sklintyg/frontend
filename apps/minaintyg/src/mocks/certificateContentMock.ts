export const certificateContentMock = [
  {
    heading: 'Smittbärarpenning',
    body: `
      <h3>Avstängning enligt smittskyddslagen på grund av smitta</h3>
      <p>Ej angivet</p>
    `,
  },
  {
    heading: 'Grund för medicinskt underlag',
    body: `
      <h3>Intyget är baserat på</h3>
      <h4>Min undersökning av patienten</h4>
      <p>2023-02-07</p>
      <h4>Min telefonkontakt med patienten</h4>
      <p>Ej angivet</p>
      <h4>Journaluppgifter från den</h4>
      <p>Ej angivet</p>
      <h4>Ange vad annat är</h4>
      <p>Ej angivet</p>
    `,
  },
  {
    heading: 'Sysselsättning',
    body: `
      <h3>I relation till vilken sysselsättning bedömer du arbetsförmågan?</h3>
      <ul>
        <li>Nuvarande arbete</li>
      </ul>
      <h3>Ange yrke och arbetsuppgifter</h3>
      <p>Kontorsarbetare</p>

      <table>
        <thead>
          <tr>
            <th>Diagnoskod enligt ICD-10 SE</th>
            <th>Diagnos</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>S829</td>
            <td>Underbensfraktur med ospecificerad lokalisation</td>
          </tr>
        </tbody>
      </table>
    `,
  },
  {
    heading: 'Sjukdomens konsekvenser',
    body: `
      <h3>Funktionsnedsättning</h3>
      <h4>Beskriv undersökningsfynd, testresultat och observationer</h4>
      <p>Patienten har en underbensfraktur. Behandlas med helbensgrips och har därför nedsatt rörlighet i benet. Patienten
        får inte stödja på det skadade benet. Går långsamt och ostadigt med kryckor. Påtalar smärta vid undersökningen.
      </p>
      <h3>Aktivitetsbegränsning</h3>
      <h4>Beskriv vad patienten inte kan göra på grund av sin sjukdom. Ange vad uppgiften grundas
        på.</h4>
      <p>Patienten kan inte gå utan hjälpmedel. Klarar inte av att köra bil. Således mycket svårt att ta sig till arbetet.
        Dessutom problem med trappor på arbetsplatsen. Patienten kan även ha svårt att fokusera på sina arbetsuppgifter
        pga smärta.</p>
    `,
  },
  {
    heading: 'Medicinsk behandling',
    body: `
      <h3>Pågående medicinska behandlingar/åtgärder</h3>
      <h4>Ange vad syftet är och om möjligt tidplan samt ansvarig vårdenhet</h4>
      <p>Erhållit helbensgips, 0-belastning med kryckor. Smärtlindring.</p>
      <h3>Planerade medicinska behandlingar/åtgärder</h3>
      <h4>Ange vad syftet är och om möjligt tidplan samt ansvarig vårdenhet</h4>
      <p>Gips totalt 4 månader, initialt helbensgips. Rtg-kontroll efter 2, 4 och 6 veckor av frakturens läge. Om stabilt
        frakturläge vid 6 v kontroll byte till underbensgips. Rehabilitering efter avslutad gipsbehandling.</p>
    `,
  },
  {
    heading: 'Bedömning',
    body: `
      <h3>Min bedömning av patientens nedsättning av arbetsförmågan</h3>
      <table>
        <thead>
          <tr>
            <th>Nedsättningsgrad</th>
            <th>Från och med</th>
            <th>Till och med</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>100 procent</td>
            <td>2023-02-07</td>
            <td>2023-02-20</td>
          </tr>
        </tbody>
      </table>

      <h3>Patientens arbetsförmåga bedöms nedsatt längre tid än den som Socialstyrelsens
        försäkringsmedicinska beslutsstöd
        anger, därför att</h3>
      <p>Ej angivet</p>

      <h3>Kommer möjligheterna till återgång i arbete försämras om arbetstiden förläggs på annat
        sätt än att arbetstiden
        minskas lika mycket varje dag?</h3>
      <p>Ej angivet</p>

      <h3>Beskriv de medicinska skälen till att möjligheterna till återgång i arbete försämras</h3>
      <p>Ej angivet</p>

      <h3>Resor till och från arbetet med annat färdmedel än normalt kan göra det möjligt för
        patienten att återgå till
        arbetet under sjukskrivningsperioden.</h3>
      <p>Ej angivet</p>

      <h3>Prognos för arbetsförmåga utifrån aktuellt undersökningstillfälle</h3>
      <p>Patienten förväntas kunna återgå helt i nuvarande sysselsättning inom 1 månad.</p>
    `,
  },
  {
    heading: 'Åtgärder',
    body: `
      <h3>Här kan du ange åtgärder som du tror skulle göra det lättare för patienten att återgå i arbete</h3>
      <ul>
        <li>Arbetsanpassning</li>
      </ul>
      <h3>Här kan du ange fler åtgärder. Du kan också beskriva hur åtgärderna kan underlätta återgång i arbete.</h3>
      <p>Möjlighet till distansarbete i hemmet skulle underlätta mycket för patienten.</p>
    `,
  },
  {
    heading: 'Övriga upplysningar',
    body: `
      <h3>Övriga upplysningar</h3>
      <p>Ej angivet</p>
    `,
  },
  {
    heading: 'Kontakt',
    body: `
      <h3>Kontakt med Försäkringskassan</h3>
      <h4>Jag önskar att Försäkringskassan kontaktar mig</h4>
      <p>Ej angivet</p>
      <h4>Ange gärna varför du vill ha kontakt</h4>
      <p>Ej angivet</p>
    `,
  },
].map(({ heading, body }) => ({ heading, body: body.trim().replace(/\s+/g, ' ') }))
