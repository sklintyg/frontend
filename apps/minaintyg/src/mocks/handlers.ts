/* eslint-disable import/no-extraneous-dependencies */
import { fakeCertificateLabel, faker, fakerFromSchema } from '@frontend/fake'
import { rest } from 'msw'
import { certificateListItemSchema } from '../schema/certificateList.schema'
import { userSchema } from '../schema/user.schema'

export const handlers = [
  rest.get('/api/user', (_, res, ctx) =>
    res(
      ctx.status(200),
      ctx.json(
        fakerFromSchema(userSchema, {
          stringMap: {
            personName: faker.name.fullName,
          },
        })()
      )
    )
  ),
  rest.post('/api/certificates', (_, res, ctx) =>
    res(
      ctx.status(200),
      ctx.json({
        content: Array.from(
          { length: 5 },
          fakerFromSchema(certificateListItemSchema, {
            stringMap: {
              title: fakeCertificateLabel,
              certificateId: faker.datatype.uuid,
              timestamp: () => faker.date.between('2021-01-01T00:00:00.000Z', new Date().toISOString()).toISOString(),
            },
          })
        ),
      })
    )
  ),
]
