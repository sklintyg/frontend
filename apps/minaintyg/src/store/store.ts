import { configureStore } from '@reduxjs/toolkit'
import { api } from './api'
import { errorMiddleware } from './middleware/error.middleware'
import { reducer } from './reducer'
import { idpMiddleware } from './slice/idpMiddleware'

export const store = configureStore({
  reducer,
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat([api.middleware, idpMiddleware, errorMiddleware]),
})

export type AppDispatch = typeof store.dispatch
