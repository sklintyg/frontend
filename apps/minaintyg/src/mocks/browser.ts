/* eslint-disable import/no-extraneous-dependencies */
import { faker } from 'fake'
import { setupWorker } from 'msw'
import { handlers } from './handlers'

faker.setLocale('sv')

export const worker = setupWorker(...handlers)
