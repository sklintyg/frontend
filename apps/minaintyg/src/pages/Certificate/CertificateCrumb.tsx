import { skipToken } from '@reduxjs/toolkit/query'
import { useGetCertificateQuery } from '../../store/api'

export function CertificateCrumb({ id }: { id: string }) {
  const { data: certificate } = useGetCertificateQuery(id ? { id } : skipToken)
  return certificate ? <>{certificate.metadata.type.name}</> : null
}
