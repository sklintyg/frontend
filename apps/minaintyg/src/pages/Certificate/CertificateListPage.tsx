import { IDSSpinner } from '@frontend/ids-react-ts'
import { useSessionStorage } from '@react-hooks-library/core'
import { SortDirection } from 'react-stately'
import { PageHeading } from '../../components/PageHeading/PageHeading'
import { PageHeadingDescription } from '../../components/PageHeading/PageHeadingDescription'
import { SupportLink } from '../../components/error/SupportLink/SupportLink'
import { TechnicalIssueAlert } from '../../components/error/TechnicalIssueAlert'
import { useGetCertificatesFilterQuery, useGetCertificatesQuery } from '../../store/api'
import { useAppSelector } from '../../store/hooks'
import { CertificateList } from './components/CertificateList'
import { CertificateListFilter } from './components/CertificateListFilter/CertificateListFilter'
import { CertificateListOrder } from './components/CertificateListOrder/CertificateListOrder'
import { EmptyCertificateListInfo } from './components/EmptyCertificateListInfo'
import {ReadMoreAboutAction} from "./components/ReadMoreAboutDialog/ReadMoreAboutAction";

export function CertificateListPage() {
  const [order, setOrder] = useSessionStorage<SortDirection>('certificate-list-order', 'descending')
  const { submitFilters } = useAppSelector((state) => state.certificateFilter)
  const { isLoading: isLoadingList, data: list, error } = useGetCertificatesQuery(submitFilters)
  const { isLoading: isLoadingFilters, data: filter } = useGetCertificatesFilterQuery()
  const isLoading = isLoadingList || isLoadingFilters

  return (
    <>
      <PageHeading heading="Intyg">
        <PageHeadingDescription>
          Här hittar du dina digitala intyg som vården skrivit. I Intyg kan du läsa, skicka och skriva ut dina intyg. Saknar du ett intyg,
          eller vill begära ett intyg? Vänd dig till din mottagning. Klicka på länken “Läs mer” nedan om du vill veta vad du kan göra i
          Intyg.
        </PageHeadingDescription>
        <ReadMoreAboutAction />
      </PageHeading>
      <CertificateListFilter listed={list?.content.length ?? 0} />
      <CertificateListOrder setOrder={setOrder} order={order} />
      {isLoading && <IDSSpinner data-testid="certificate-list-spinner" />}
      {error && (
        <TechnicalIssueAlert error={error}>
          <p>
            Intygen kunde inte visas på grund av ett tekniskt fel. Försök igen senare. Om felet kvarstår kontakta <SupportLink />.
          </p>
          <p>Om du inte kan nå ditt intyg, kontakta din läkare för att få en kopia av intyget.</p>
        </TechnicalIssueAlert>
      )}
      {list && list.content.length === 0 && <EmptyCertificateListInfo total={filter?.total ?? 0} />}
      {list && <CertificateList certificates={list.content} order={order} />}
    </>
  )
}
