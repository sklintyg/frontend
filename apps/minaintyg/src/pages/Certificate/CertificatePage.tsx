import { IDSSpinner } from '@frontend/ids-react-ts'
import { skipToken } from '@reduxjs/toolkit/query'
import { useParams } from 'react-router-dom'
import { PageDivider } from '../../components/PageDivider/PageDivider'
import { PageHeading } from '../../components/PageHeading/PageHeading'
import { PageHeadingDescription } from '../../components/PageHeading/PageHeadingDescription'
import { useGetCertificateQuery } from '../../store/api'
import { CertificateActions } from './components/CertificateActions/CertificateActions'
import { CertificateBody } from './components/CertificateBody/CertificateBody'
import { CertificateEventsInfo } from './components/CertificateEventsInfo/CertificateEventsInfo'
import { CertificateFooter } from './components/CertificateFooter'
import { CertificateInformation } from './components/CertificateInformation'
import { CertificateReplacedAlert } from './components/CertificateReplacedAlert'
import { CertificateStatusBadge } from './components/CertificateStatusBadge'
import { ReadCertificateError } from './components/ReadCertificateError'
import { CertificateAttentionAlert } from './components/CertificateAttentionAlert/CertificateAttentionAlert'
import { CertificatePreambleText } from './components/CertificatePremableText/CertificatePreambleText'
import { ReadMoreAboutAction } from './components/ReadMoreAboutDialog/ReadMoreAboutAction'

const FALLBACK_DESCRIPTION = `Det här är ditt intyg. Intyget innehåller all information vården fyllt i. Du kan inte ändra något i ditt intyg. Har du frågor kontaktar du den som skrivit ditt intyg.`

export function CertificatePage() {
  const { id } = useParams()
  const { data: certificate, isLoading, error } = useGetCertificateQuery(id ? { id } : skipToken)

  return (
    <>
      <PageHeading heading="Läs och hantera ditt intyg">
        {certificate && (
          <>
            <CertificatePreambleText texts={certificate.texts} />
            {certificate.metadata.statuses.includes('REPLACED') && <CertificateReplacedAlert />}
          </>
        )}
        {error && <PageHeadingDescription>{FALLBACK_DESCRIPTION}</PageHeadingDescription>}
        <ReadMoreAboutAction />
      </PageHeading>

      {isLoading && <IDSSpinner data-testid="spinner" />}
      {error && <ReadCertificateError id={id} error={error} />}
      {certificate && (
        <>
          <div className="mb-5">
            <CertificateAttentionAlert availableFunctions={certificate.availableFunctions} />
            <CertificateActions
              recipient={certificate.metadata.recipient}
              availableFunctions={certificate.availableFunctions}
              id={certificate.metadata.id}
            />
          </div>

          <div className="md:hidden">
            <PageDivider />
          </div>

          <div className="rounded-[10px] border-stone-line md:border md:p-7 md:shadow-[0_0_2px_0px_rgba(0,0,0,0.3)]">
            <div className="mb-5 flex flex-col justify-between gap-2.5 md:flex-row md:gap-5">
              <h2 className="ids-heading-2">{certificate.metadata.type.name}</h2>
              <div className="flex gap-1">
                {certificate.metadata.statuses.map((status) => (
                  <CertificateStatusBadge key={status} status={status} />
                ))}
              </div>
            </div>
            <div className="mb-5">
              <CertificateInformation {...certificate.metadata} />
            </div>
            <h3 className="ids-heading-4 mb-0">Händelser</h3>
            <CertificateEventsInfo events={certificate.metadata.events} />

            <article className="ids-certificate">
              <CertificateBody content={certificate.content} />
              <CertificateFooter {...certificate.metadata} />
            </article>
          </div>
        </>
      )}
    </>
  )
}
