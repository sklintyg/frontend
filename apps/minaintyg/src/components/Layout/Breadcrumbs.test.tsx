import { fakerFromSchema } from '@frontend/fake'
import { render, screen } from '@testing-library/react'
import { rest } from 'msw'
import { Provider } from 'react-redux'
import { Outlet, Route, RouterProvider, createMemoryRouter, createRoutesFromChildren } from 'react-router-dom'
import { server } from '../../mocks/server'
import { CertificateCrumb } from '../../pages/Certificate/CertificateCrumb'
import { certificateSchema } from '../../schema/certificate.schema'
import { store } from '../../store/store'
import { Breadcrumbs } from './Breadcrumbs'

const renderComponent = (initialEntries = ['/']) =>
  render(
    <Provider store={store}>
      <RouterProvider
        router={createMemoryRouter(
          createRoutesFromChildren([
            <Route
              key="root"
              path="/"
              handle={{ crumb: () => 'Start' }}
              element={
                <div>
                  <Breadcrumbs />
                  <Outlet />
                </div>
              }
            >
              <Route index element={<p>Start</p>} />
              <Route path="/intyg" handle={{ crumb: () => 'Intyg' }}>
                <Route index element={<p>Certificates List</p>} />
                <Route
                  path=":id"
                  handle={{
                    crumb: ({ id }: { id: string }) => <CertificateCrumb id={id} />,
                  }}
                  element={<p>Intyget</p>}
                />
              </Route>
            </Route>,
          ]),
          { initialEntries }
        )}
      />
    </Provider>
  )

it('Should render as expected with one level', () => {
  const { baseElement } = renderComponent(['/'])
  expect(baseElement).toMatchSnapshot()
})

it('Should render as expected with two level', () => {
  const { baseElement } = renderComponent(['/intyg'])
  expect(baseElement).toMatchSnapshot()
})

it('Should render as expected with three levels', async () => {
  const certificateName = 'läkares anmälan till transportstyrelsen'
  server.use(
    rest.get('/api/certificate/:id', (_, res, ctx) =>
      res(
        ctx.status(200),
        ctx.json(
          fakerFromSchema(certificateSchema)({
            metadata: {
              type: {
                name: certificateName,
              },
            },
          })
        )
      )
    )
  )
  const { baseElement } = renderComponent(['/intyg/12345'])
  expect(await screen.findByText(certificateName)).toBeInTheDocument()
  expect(baseElement).toMatchSnapshot()
})
