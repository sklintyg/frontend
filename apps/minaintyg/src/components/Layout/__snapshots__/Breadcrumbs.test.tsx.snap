// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Should render as expected with one level 1`] = `
<div>
  <div>
    <div
      class="mb-5"
    >
      <ids-breadcrumbs>
        <span>
          Start
        </span>
      </ids-breadcrumbs>
    </div>
    <p>
      Start
    </p>
  </div>
</div>
`;

exports[`Should render as expected with three levels 1`] = `
<div>
  <div>
    <div
      class="mb-5"
    >
      <ids-breadcrumbs>
        <ids-crumb>
          <a
            href="https://e-tjanster.1177.se/mvk/"
          >
            Start
          </a>
        </ids-crumb>
        <ids-crumb>
          <a
            href="/intyg"
          >
            Intyg
          </a>
        </ids-crumb>
        <span>
          läkares anmälan till transportstyrelsen
        </span>
        <ids-crumb
          slot="mobile"
        >
          <a
            class="no-underline"
            href="/intyg"
          >
            Intyg
          </a>
        </ids-crumb>
      </ids-breadcrumbs>
    </div>
    <p>
      Intyget
    </p>
  </div>
</div>
`;

exports[`Should render as expected with two level 1`] = `
<div>
  <div>
    <div
      class="mb-5"
    >
      <ids-breadcrumbs>
        <ids-crumb>
          <a
            href="https://e-tjanster.1177.se/mvk/"
          >
            Start
          </a>
        </ids-crumb>
        <span>
          Intyg
        </span>
        <ids-crumb>
          <a
            class="no-underline"
            href="https://e-tjanster.1177.se/mvk/"
          >
            Start
          </a>
        </ids-crumb>
      </ids-breadcrumbs>
    </div>
    <p>
      Certificates List
    </p>
  </div>
</div>
`;
